source/puppet/4.4/reference/static_catalogs.md

```diff
@@ -0,0 +1,132 @@
+
+Puppet Server locates these commands via the `code-id-command` and `code-content-command` settings in Puppet Server's [`puppetserver.conf`][] file. Puppet Server runs the `code-id-command` each time it compiles a static catalog, and it runs the `code-content-command` each time an agent requests file contents from the `static_file_content` endpoint.
+
+> **Note:** The Puppet Server process must be able to execute these scripts. Puppet Server also validates their output and checks their exit codes. Environment names can contain only alphanumeric characters and underscores (`_`). The `code_id` can  contain only alphanumeric characters and dashes (`-`), underscores (`_`), semicolons (`;`), and colons (`:`). If either command returns a non-zero exit code, Puppet Server logs an error and returns the error message and a 500 response code to the API request.
+
+Puppet Server validates the standard output of each of these scripts, and if the output's acceptable, it adds the results to the catalog as their respective parameters' values. This lets you use any versioning or synchronization tools you want, as long as you can write scripts that produce a valid string for the `code_id` and code content using the catalog's `code_id` and file's environment.
+
+The `code-id-command` and `code-content-command` scripts can be as simple or complex as necessary.
+
+> **Note:** The following examples demonstrate how Puppet Server passes arguments to the `code-id-command` and `code-content-command` scripts and how Puppet Server uses the results to return a specific version of a file resource. The scripts themselves are not recommended for production use.
+
+For instance, if files in an environment are simply managed by git, a `code-id-command` script can be nothing more than this, with the environment name passed to the script as the first command-line argument:
+
+``` bash
+#!/bin/bash
+set -e
+if [[ -z $1 ]]; then
+  echo Expected an environment >&2
+  exit 1
+fi
+cd /etc/puppetlabs/code/environments/"$1" && git rev-parse HEAD
+```
+
+As long as the script's exit code is zero, Puppet Server uses the script's standard output as the catalog's `code_id`.
+
+A `code-content-command` script can also be simple. Puppet Server passes the environment name as the first command-line argument, the `code_id` as the second argument, and the path to the file resource from its `content_uri` as the third argument:
+
+``` bash
+#!/bin/bash
+set -e
+if [[ $# < 3 ]]; then
+  echo Expected environment, code-id, file-path >&2
+  exit 1
+fi
+cd /etc/puppetlabs/code/environments/"$1" && git show $2:$3
```

Comment:
`$3` should, because the path may contain spaces or other special chars. It couldn't hurt to quote `$2` as well IMO.